actuator-firmware/src/pwm.h

```diff
@@ -8,17 +8,32 @@ extern "C" {
 #include <ch.h>
 #include <hal.h>
 
-enum pwm_channel {
-    PWM_CHANNEL_0, // Timer 16 channel 0
-    PWM_CHANNEL_1, // Timer 17 channel 0
-    PWM_CHANNEL_2, // Timer 1 channel 1
-    PWM_CHANNEL_3, // Timer 1 channel 2
-};
+/// Mask to select channel number, 1 indexed
+#define PWM_CHANNEL(x) (1 << (x - 1))
 
-void pwm_init(uint32_t frequency, uint32_t period);
+/** Low level init of PWM channels for a timer.
+ *
+ * @param [in] tim Address of the timer to use, example STM32_TIM1.
+ * @param [in] channel_mask Which channels to enable, OR'd together. You can
+ * use PWM_CHANNEL(n) to enable channel N.
+ * @param [in] clock_frequency The frequency of the clock running this
+ * particular timer peripheral. You can use macros such as STM32_TIMCLK2 to
+ * compute it from the RCC configuration.
+ * @param [in] timer_frequency The frequency at which the timer should count.
+ * @param [in] timer_period The timer period.
+ */
+void pwm_setup_channels(stm32_tim_t* tim,
+                        uint8_t channel_mask,
+                        uint32_t clock_frequency,
+                        uint32_t timer_frequency,
+                        uint32_t timer_period);
 
-/* Set given pulse width in clock ticks to the corresponding PWM channel */
-void pwm_set_pulsewidth(enum pwm_channel channel, uint32_t pulsewidth);
+/** Sets the pulse width for that particular timer and channel.
+ *
+ * Note that the channel number is 0-indexed here, while in the datasheet it is
+ * 1-indexed.
+ */
+void pwm_set_pulsewidth(stm32_tim_t* tim, uint8_t channel, uint32_t width);
```

Comment:
Maybe we want also an API with on/off and configurable pwm for those

Reply:
What would you want to configure ?

Reply:
on/off pulsewidth

Reply:
Well the on pulse width is specified from the duty cycle, so once per channel (hence this function). On the other hand, the cycle duration can only be specified on a per-timer basis (i.e. for all channels at once), and this is why it is provided as a parameter to `pwm_setup_channels`.